### PR TITLE
Ability to copy a layer (Issue 2506)

### LINKED ISSUE #2506: Ability to copy layer
# Feature Request

**Describe the solution you'd like**
In regards to split map functionality would be good to be able to copy existing layer and then set different dimensions for that copy. To compare the same layer but in different time. 
Probably the best place for the copy button would be in layer editor before Delete.

### PR DESCRIPTION
**Pull request by DailisLangovskis**
## Description

Allows the user to create a copy of a layer from layer editor panel using a button and confirming the action.

## Related issues or pull requests

closes #2506 

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Dependency updates
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe)
- [ ] I am unsure (we'll look into it together)

## Do you introduce a breaking change?

- [ ] Yes
- [x] No
- [ ] I am unsure (no worries, we'll find out)

## Checklist

- [x] I understand and agree that the changes in this PR will be licensed under the [MIT License]
- [x] I have followed the [guidelines for contributing](https://github.com/hslayers/hslayers-ng/blob/master/CONTRIBUTING.md)
- [x] The proposed change fits to the content of the [code of conduct](https://github.com/hslayers/hslayers-ng/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have added or updated tests and documentation, and the test suite passes (run `npm test` locally)
- [ ] I'm lost; why do I have to check so many boxes? Please help!

### CODE REVIEW

**Comment by jmacura** [comment]
A nice enhancement to this would be: In the confirmation dialog, instead of simply asking for confirmation, ask for a new name for the copy. What do you think?

File: projects/hslayers/src/components/layermanager/layermanager.service.ts [hunk 1]
```diff
@@ -1175,4 +1176,25 @@ export class HsLayerManagerService {
       window.URL.revokeObjectURL(url);
     }, 0);
   }
+
+  /*
+    Creats a copy of the currentLayer
+  */
+  copyLayer(newTitle: string): void {
+    const copiedLayer = new VectorLayer(
+      this.currentLayer.layer.getProperties()
```
Comment (raitisbe) [inline projects/hslayers/src/components/layermanager/layermanager.service.ts:1185]: Will need to also clone the layers Source. Currently the cloned layer and original layer shares the same source object, which might create problems - changing dimension or other properties in one changes other.

**Comment by raitisbe** [comment]
I think the Object.create() approach might create problems after all. I tried using that approach when cloning maps View and had unintended consequences - ol_uid was the same for the original and clone and the properties on clone such as center/zoom were still in sync with the original view when moving the map. I expect something similar might happen also with layer sources. Need to test more.

Bulletproof but a lengthy way would be to recreate the sources, copy getProperties() and serialize/deserialize features (maybe can also insert feature instances directly into the new source and they would be copied not referenced  - have to check) in case of vectorSource or copy the current layer params in case of WMS. `Object.create` sure looks tempting considering this.

**Comment by DailisLangovskis** [comment]
> I think the Object.create() approach might create problems after all. I tried using that approach when cloning maps View and had unintended consequences - ol_uid was the same for the original and clone and the properties on clone such as center/zoom were still in sync with the original view when moving the map. I expect something similar might happen also with layer sources. Need to test more.
> 
> Bulletproof but a lengthy way would be to recreate the sources, copy getProperties() and serialize/deserialize features (maybe can also insert feature instances directly into the new source and they would be copied not referenced - have to check) in case of vectorSource or copy the current layer params in case of WMS. `Object.create` sure looks tempting considering this.

I will try to recreate the sources for each layer type

**Comment by raitisbe** [comment]
Will need to get rid of owsFilling subject and rewrite to promises at some point

**Review by raitisbe — APPROVED** [review]

**Pull request merged**